api/proto/teleport/hardwarekeyagent/v1/hardwarekeyagent_service.proto

```diff
@@ -0,0 +1,129 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+syntax = "proto3";
+
+package teleport.hardwarekeyagent.v1;
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/hardwarekeyagent/v1;hardwarekeyagentv1";
+
+// HardwareKeyAgentService provides an agent service for hardware key (PIV) signatures.
+// This allows multiple Teleport clients to share a PIV connection rather than blocking
+// each other, due to the exclusive nature of PIV connections. This also enabled shared
+// hardware key states, such as a custom PIN cache shared across Teleport clients.
+service HardwareKeyAgentService {
+  // Ping the agent service to check if it is active.
+  rpc Ping(PingRequest) returns (PingResponse) {}
+  // Sign produces a signature with the provided options for the specified hardware private key
+  //
+  // This rpc implements Go's crypto.Signer interface.
+  rpc Sign(SignRequest) returns (Signature) {}
+}
+
+// PingRequest is a request to Ping.
+message PingRequest {}
+
+// PingResponse is a response to Ping.
+message PingResponse {
+  // PID is the PID of the client process running the agent.
+  uint32 pid = 1;
+}
+
+// SignRequest is a request to perform a signature with a specific hardware private key.
+message SignRequest {
+  // Digest is a hashed message to sign.
+  bytes digest = 1;
+  // Hash is the hash function used to prepare the digest.
+  Hash hash = 2;
+  // SaltLength specifies the length of the salt added to the digest before a signature.
+  // This salt length is precomputed by the client, following the crypto/rsa implementation.
+  // Only used, and required, for PSS RSA signatures.
+  uint32 salt_length = 3;
+  // KeyRef references a specific hardware private key.
+  KeyRef key_ref = 4;
+  // KeyInfo contains additional, optional key info which generally will improve UX by
+  // giving the agent context about the key, such as whether PIN/touch prompts are
+  // expected, or what cluster login is trying to interface with the key.
+  KeyInfo key_info = 5;
+  // Command is the client command or action requiring a signature, e.g. "tsh ssh server01".
+  // The agent can include this detail in PIN/touch prompts to show the origin of the
+  // signature request to the user.
+  string command = 6;
+}
+
+// Signature is a private key signature.
+message Signature {
+  // For an RSA key, signature should be either a PKCS #1 v1.5 or PSS signature,
+  // depending on the hash and salt chosen. For an (EC)DSA key, it should be a
```

Comment:
Would there be any benefit in also including the type of key in the message so that it's easier for the client to know how to parse it?

Reply:
Not with go's crypto library in mind - the two types of signatures are handled interchangeably and handled by the caller. This comment is actually paraphrased from the `crypto.Signer` interface:

```
// Sign signs digest with the private key, possibly using entropy from
// rand. For an RSA key, the resulting signature should be either a
// PKCS #1 v1.5 or PSS signature (as indicated by opts). For an (EC)DSA
// key, it should be a DER-serialised, ASN.1 signature structure.
// ...
Sign(rand io.Reader, digest []byte, opts SignerOpts) (signature []byte, err error)
```